src/sst/core/serialization/impl/serialize_array.h

```diff
@@ -20,200 +20,197 @@
+            }
+            [[fallthrough]];
+
+        default:
+            if constexpr ( is_trivial_element_v<ELEM_T> )
+                ser.raw(&(*aPtr)[0], sizeof(ELEM_T) * SIZE);
+            else
+                serialize_array(ser, &(*aPtr)[0], SIZE, serialize_array_element<ELEM_T>);
+            break;
         }
-
-        // TODO: Implement mapping mode
     }
 };
 
-/***  For dynamically allocated arrays ***/
+} // namespace pvt
 
-/**
-   Version of serialize that works for dynamically allocated arrays of
-   arithmetic types and enums.
- */
-template <class T, class IntType>
-class serialize_impl<pvt::ser_array_wrapper<T, IntType>, std::enable_if_t<std::is_arithmetic_v<T> || std::is_enum_v<T>>>
-{
-    template <class A>
-    friend class serialize;
-    void operator()(pvt::ser_array_wrapper<T, IntType> arr, serializer& ser)
-    {
-        ser.binary(arr.bufptr, arr.sizeptr);
+// Serialize fixed arrays and pointers to them
+template <typename ELEM_T, size_t SIZE>
+struct serialize_impl<ELEM_T[SIZE]> : pvt::serialize_impl_fixed_array<ELEM_T[SIZE], ELEM_T, SIZE>
+{};
 
-        // TODO: Implement mapping mode
-    }
-};
+template <typename ELEM_T, size_t SIZE>
+struct serialize_impl<std::array<ELEM_T, SIZE>> :
+    pvt::serialize_impl_fixed_array<std::array<ELEM_T, SIZE>, ELEM_T, SIZE>
+{};
 
-/**
-   Version of serialize that works for dynamically allocated arrays of
-   non base types.
- */
-template <class T, class IntType>
-class serialize_impl<
-    pvt::ser_array_wrapper<T, IntType>, std::enable_if_t<!std::is_arithmetic_v<T> && !std::is_enum_v<T>>>
-{
-    template <class A>
-    friend class serialize;
-    void operator()(pvt::ser_array_wrapper<T, IntType> arr, serializer& ser)
-    {
-        ser.primitive(arr.sizeptr);
-        for ( int i = 0; i < arr.sizeptr; i++ ) {
-            ser& arr[i];
-        }
+template <typename ELEM_T, size_t SIZE>
+struct serialize_impl<ELEM_T (*)[SIZE]> : pvt::serialize_impl_fixed_array<ELEM_T (*)[SIZE], ELEM_T, SIZE>
+{};
 
-        // TODO: Implement mapping mode
-    }
-};
+template <typename ELEM_T, size_t SIZE>
+struct serialize_impl<std::array<ELEM_T, SIZE>*> :
+    pvt::serialize_impl_fixed_array<std::array<ELEM_T, SIZE>*, ELEM_T, SIZE>
+{};
 
-/**
-   Version of serialize that works for statically allocated arrays of
-   void*.
- */
-template <class IntType>
-class serialize_impl<pvt::ser_array_wrapper<void, IntType>>
+// Serialize dynamic arrays
+template <typename ELEM_T, typename SIZE_T>
+struct serialize_impl<pvt::array_wrapper<ELEM_T, SIZE_T>>
 {
-    template <class A>
-    friend class serialize;
-    void operator()(pvt::ser_array_wrapper<void, IntType> arr, serializer& ser)
+    void operator()(pvt::array_wrapper<ELEM_T, SIZE_T>& ary, serializer& ser)
     {
-        ser.binary(arr.bufptr, arr.sizeptr);
-
-        // TODO: Implement mapping mode
+        switch ( const auto mode = ser.mode() ) {
+        case serializer::MAP:
```

Comment:
Can you add this mode?  I know of several dynamically allocated arrays that will need this feature immediately.

Reply:
I can add mapping mode but I was waiting to see how the `ObjectMap` should be implemented. We have a container `ObjectMap` -- should arrays just use `ObjectMapContainer` or create a `ObjectMapArray` class which inherits from `ObjectMapContainer`?  How will we store the array size in it, and how does the size stored change w.r.t. serialization (I mean, how to keep the `ObjectMapArray` class or whatever in-sync with the array address and size in the program? Or do we only need the `ObjectMapArray` to live the life of a single serialization call and then it can be destroyed and none of the array or other information stored in it matters antmore?).